### PR TITLE
Swap dimensions when image is rotated on native devices

### PR DESCRIPTION
**Pull request by francoisl**
### Details
- Fix attachment previews that look "small" (don't use all the available space) on Android.

### Fixed Issues
$ https://github.com/Expensify/App/issues/13300  

### Tests
On native devices (Android, iOS, mWeb Chrome & Safari):
1. Open any chat
2. In the compose box, click the `+` button > `📎  Add attachment` > `📸 Take photo`
3. Take a photo in portrait mode
4. On the "Send attachment" preview page, make sure the image uses all the available width space and isn't unnecessarily squished
5. Repeat with a photo taken in landscape mode, make sure the preview also doesn't look squished.
6. Download the following two images and drag-and-drop them in any chat
208204571-007b5229-cc34-410a-81df-0c4dc52faa64&nbsp;&nbsp;&nbsp;
7. After the images are added to the chat, open them back by clicking on them
8. Make sure the images don't look unnecessarily sized down

- [x] Verify that no errors appear in the JS console

### Offline tests
The same steps below apply while offline - we allow taking a photo as an attachment while offline.

### QA Steps
On native devices (Android, iOS, mWeb Chrome & Safari):
1. Open any chat
2. In the compose box, click the `+` button > `📎  Add attachment` > `📸 Take photo`
3. Take a photo in portrait mode
4. On the "Send attachment" preview page, make sure the image uses all the available width space and isn't unnecessarily squished
5. Repeat with a photo taken in landscape mode, make sure the preview also doesn't look squished.
6. Download the following two images and drag-and-drop them in any chat
208204571-007b5229-cc34-410a-81df-0c4dc52faa64&nbsp;&nbsp;&nbsp;
7. After the images are added to the chat, open them back by clicking on them
8. Make sure the images don't look unnecessarily sized down

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<table>
<tr>
	<td>Portrait photo
	<td>Landscape photo
<tr>
	<td>image
	<td>image
</table>

</details>

<details>
<summary>Mobile Web - Safari</summary>

image

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

image

</details>

<details>
<summary>Android</summary>

<table>
<tr>
	<td>Device
	<td>Portrait photo
	<td>Landscape photo
	<td>Wide image from a chat
	<td>Tall image from a chat
<tr>
	<td>Pixel 3a AVD
	<td>Screenshot 2022-12-13 at 4 45 36 PM
	<td>Screenshot 2022-12-13 at 4 45 53 PM
	<td>image
	<td>image
<tr>
	<td>Pixel 6 Pro AVD
	<td>
Screenshot 2022-12-13 at 4 46 12 PM
	<td>Screenshot 2022-12-13 at 4 46 37 PM
	<td>image
	<td>image
<tr>
	<td>Pixel 6 (physical)
	<td>
	<td>
	<td>
	<td>
</table>

</details>

### CODE REVIEW

**Comment by youssef-lr** [comment]
@francoisl I would like to test this but I don't have access to an Android device. @thesahindia Are you able to test this on a physical device?

**Comment by thesahindia** [comment]
Yeah, I can test this on a physical Android device. Looking into it in a few minutes...

**Comment by thesahindia** [comment]
I have tried multiple devices but I couldn't repro this issue. I also couldn't repro this at `Pixel 6 Pro AVD` and `Pixel 3a AVD`
Screenshot 2022-12-14 at 3 54 18 PM

**Comment by francoisl** [comment]
I wasn't able to reproduce on any AVD either (see [this comment](https://github.com/Expensify/App/issues/13300#issuecomment-1350153224) for context), only on a physical Pixel 6. Out of curiosity, can you log the `rotation` parameter on devices where you can't repro? I bet it will be `0`, and you will have `height > width`.

**Comment by thesahindia** [comment]
> I wasn't able to reproduce on any AVD either (see https://github.com/Expensify/App/issues/13300#issuecomment-1350153224 for context), only on a physical Pixel 6.

Gotcha! In that case I think I should just verify that everything works correctly on all platforms

> I bet it will be 0, and you will have height > width

Yeah, that is correct

**Comment by thesahindia** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
Screenshot 2022-12-15 at 2 33 25 AM



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>
Screenshot 2022-12-15 at 2 30 41 AM
Screenshot 2022-12-15 at 2 31 17 AM



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>
Screenshot 2022-12-15 at 2 27 29 AM
Screenshot 2022-12-15 at 2 31 45 AM




</details>

<details>


<summary>Android</summary>
Screenshot 2022-12-15 at 1 56 05 AM
Screenshot 2022-12-15 at 1 55 20 AM


</details>

**Review by thesahindia — DISMISSED** [review]
Looks good! Didn't see any regression

C+ reviewed 🎀👀🎀

cc: @youssef-lr

**Review by aldo-expensify — COMMENTED** [review]
Could it be that his only work in some Android devices? I tested in a Xiaomi A2 native with a portrait image and I get these values:

rotation: 0
imageWidth: 4000
imageHeight: 3000

The width and height are flipped since the image is taller than wider, but rotation is 0 :s

Screenshot_20221215-162934


I also couldn't find the documentation for `ImageSize.getSize` where they describe "rotation". Do you have a link? 🙏

**Comment by francoisl** [comment]
And it's not really "documentation" per se, in that the code doesn't have comments explaining the logic. It looks somewhat similar to the second answer in [this SO question](https://stackoverflow.com/questions/40910720/how-to-check-whether-the-image-is-portrait-or-landscape-in-android) though.

**Comment by aldo-expensify** [comment]
> Yes, it's [here](https://github.com/eXist-FraGGer/react-native-image-size/blob/19db01ba1eed65314be0090078048f52eff89cfe/android/src/main/java/com/existfragger/rnimagesize/RNImageSizeModule.java#L57) in the original repo. We have our own [fork](https://github.com/Expensify/react-native-image-size) of that repo, but we didn't change that logic. The rotation param is determined from the Exif properties of the image

Thanks! I guess that your condition should also consider 270 degrees, right?

> Oh man, that sucks. We could remove the rotation === 90 condition, but I'd need to check if that affects pictures you take in landscape mode then...

I'll test again with pictures taken in another device to see if those have the rotation info. Also, in the test I did, I attached the pictures directly for the camera, maybe I get something different if I take the pictures and then attach.

File: src/components/ImageView/index.native.js [hunk 1]
```diff
@@ -66,12 +66,18 @@ class ImageView extends PureComponent {
...
-        ImageSize.getSize(this.props.url).then(({width, height}) => {
+        ImageSize.getSize(this.props.url).then(({width, height, rotation}) => {
             let imageWidth = width;
             let imageHeight = height;
             const containerWidth = Math.round(this.props.windowWidth);
             const containerHeight = Math.round(this.state.containerHeight);
 
+            // On specific Android devices, the dimensions are sometimes returned to us flipped here, with `rotation` set to 90 degrees.
+            // Swap them back to make sure the image fits nicely in the container. On iOS, the rotation is always undefined, so this does not apply.
+            if (rotation === 90 && imageWidth > imageHeight) {
```
Comment (aldo-expensify) [inline src/components/ImageView/index.native.js:77]:
```suggestion
            if ((rotation === 90 || rotation === 270)  && imageWidth > imageHeight) {
```

**Review by aldo-expensify — CHANGES_REQUESTED** [review]
Retested doing:

1. `+ > Add attachment > Take photo`
2. `+ > Add attachment > Choose from galley`

In both cases when I get the preview, before finally uploading, I get the rotation correctly 👍 

**NAB:** I think I know what the problem is. The preview when you are attaching is working fine with your change, the rotation is there. The problem is after you get the image back from the server. The image from the server seems to have lost the rotation information. (NAB because I think this has to be fixed in the server)

File: src/components/ImageView/index.native.js (hunk 1 above)
Comment (francoisl) [reply src/components/ImageView/index.native.js:77]: Makes sense, though does this also fix the situation from your review [here](https://github.com/Expensify/App/pull/13575#pullrequestreview-1220127578)? Because you were getting `rotation = 0`, right?

Comment (aldo-expensify) [reply src/components/ImageView/index.native.js:77]: hehe, now I see that my message was a bit confusing [here](https://github.com/Expensify/App/pull/13575#issuecomment-1354058005).

Adding 270 is not related to the case where I get `rotation = 0`, it is just for the case when I did get the rotation and the device had been rotate in 270 degrees.

I noticed that you get `rotation = 0` if you do this:

1. Open any chat
2. In the compose box, click the + button > 📎 Add attachment > 📸 Take photo
3. Take a photo in portrait mode
4. On the "Send attachment" preview page, make sure the image uses all the available width space and isn't unnecessarily squished
5. Click "Send attachment" and wait for it to upload
6. Click the attachment thumbnail: it opens the same modal the preview uses, but now `rotation = 0`.

I think this is because when our server processes the image, it looses the "rotation" information, so I was thinking that this part may need a fix in the backend

Comment (francoisl) [reply src/components/ImageView/index.native.js:77]: Got it. Thanks for the clarification. Honestly I think we should just remove the `rotation` param altogether and swap the dimensions. I tested locally and it also prevents the issue you describe above with opening an image that was already uploaded. It also works with landscape images because we keep the aspect ratio of the image.

**Comment by francoisl** [comment]
Updated and retested, any chance you can take a look again please @aldo-expensify ?

**Review by aldo-expensify — CHANGES_REQUESTED** [review]
I think you will have to take the rotation in consideration, otherwise you will get it wrong for images that are really wider than taller.

Try this image:

Screen Shot 2022-12-16 at 1 34 54 PM

<table>
<tr>
<td>Before PR</td>
<td>after PR</td>
</tr>
<tr>
<td>

</td>
<td>



</td>
</tr>
</table>

**Comment by francoisl** [comment]
Oh, right - the original image's height is less than the `containerWidth` in this case, so the dimension swap messes things up. It seems we have two possibilities then:

1. Use the `rotation` param, and swap the width/height if `rotation` is 90 or 270.
     - This still causes an issue when you open an image whose width and height are returned swapped and `rotation` is 0 (e.g. you open an image that already exists in the chat)
     - This also means we'd need to look into why the image is losing is `rotation` Exif data once it's attached to the chat (maybe it's simply because we strip the Exif data when we upload the file?)


2. Alternative solution I just thought of: if we see that the computed `imageWidth` is less than the `containerWidth`, then we upsize the image to `containerWidth`. In other words, add the following below the `imageWidth *= aspectRatio;` line:

```js
if (imageWidth < containerWidth) {
    imageHeight = imageWidth / imageHeight * containerWidth;
    imageWidth = containerWidth;
}
```

I tested the second solution on all possible cases we've seen so far. Thougts?

**Comment by aldo-expensify** [comment]
> 2. Alternative solution I just thought of: if we see that the computed `imageWidth` is less than the `containerWidth`, then we upsize the image to `containerWidth`. In other words, add the following below the `imageWidth *= aspectRatio;` line:
> 
> ```js
> if (imageWidth < containerWidth) {
>     imageHeight = imageWidth / imageHeight * containerWidth;
>     imageWidth = containerWidth;
> }
> ```
> 
> I tested the second solution on all possible cases we've seen so far. Thougts?

This change makes it work fine for the testing image [here](https://github.com/Expensify/App/pull/13575#pullrequestreview-1221506815) (much wider than taller image), but I don't see it working for a portrait image taken with the phone (original issue) mainly because:

```
imageWidth: 4000
containerWidth: 360
imageHeight: 3000
```

The `imageWidth` of a picture taken of the phone is much larger than `containerWidth`, so you will never enter in your if

**Comment by francoisl** [comment]
We still need the width/height swap though, did you remove it? Here's the final code I'm testing with:

```js
ImageSize.getSize(this.props.url).then(({width, height}) => {
            let imageWidth = width;
            let imageHeight = height;
            console.info('Original dimensions', {imageWidth, imageHeight});
            const containerWidth = Math.round(this.props.windowWidth);
            const containerHeight = Math.round(this.state.containerHeight);

            // On certain Android devices, the dimensions are sometimes returned to us flipped here. Swap them to make sure
            // the image uses all the container space available.
            if (imageWidth > imageHeight) {
                [imageWidth, imageHeight] = [imageHeight, imageWidth];
            }

            const aspectRatio = Math.min(containerHeight / imageHeight, containerWidth / imageWidth);

            imageHeight = Math.round(aspectRatio * imageHeight);
            imageWidth = Math.round(aspectRatio * imageWidth);

            if (imageWidth < containerWidth) {
                imageHeight = imageWidth / imageHeight * containerWidth;
                imageWidth = containerWidth;
                console.info('Upsized image', {imageWidth, imageHeight});
            }

            // Resize the image to max dimensions possible on the Native platforms to prevent crashes on Android. To keep the same behavior, apply to IOS as well.
            const maxDimensionsScale = 11;
            imageHeight = Math.min(imageHeight, (this.props.windowHeight * maxDimensionsScale));
            imageWidth = Math.min(imageWidth, (this.props.windowWidth * maxDimensionsScale));
            console.info('Final dimensions', {imageWidth, imageHeight, containerWidth, containerHeight});
            this.setState({imageHeight, imageWidth});
        });
```

and it still resolves the original issue. Are you seeing different results?

**Comment by aldo-expensify** [comment]
ahh, I didn't read correctly your [initial comment](https://github.com/Expensify/App/pull/13575#issuecomment-1355757943), I changed my code incorrectly. Let me try again.

**Comment by aldo-expensify** [comment]
Cases:
1. 🟢 Original issue case: portrait photo)
2. 🟢 Wider than taller screenshot 
image
3. 🔴 Taller than wider screenshot:

Screen Shot 2022-12-16 at 1 35 53 PM

<table>
<tr>
<td>Before PR</td>
<td>after PR</td>
</tr>
<tr>
<td>

</td>
<td>

</td>
</tr>
</table>

The cases 2. and 3. have the correct original height and width, no need to flip them, and your code is rotating in these cases when it shouldn't. I'm starting to think that you can't really resolve this without the missing `orientation` information, because:
- You only need to flip height and width when it comes incorrectly flipped
- You can only know if it is incorrectly flipped using the `rotation` information

**Comment by francoisl** [comment]
Yeah I think you're right. We can fix your issue above by resizing the image a second time to fit the container height, but it inevitably breaks another case where we're resizing when we shouldn't, etc. Let's go with the `rotation` param for now, and look into why it's stripped out in the image stored in our backend as a follow-up.

File: src/components/ImageView/index.native.js [hunk 2]
```diff
@@ -66,12 +66,18 @@ class ImageView extends PureComponent {
...
-        ImageSize.getSize(this.props.url).then(({width, height}) => {
+        ImageSize.getSize(this.props.url).then(({width, height, rotation}) => {
             let imageWidth = width;
             let imageHeight = height;
             const containerWidth = Math.round(this.props.windowWidth);
             const containerHeight = Math.round(this.state.containerHeight);
 
+            // On specific Android devices, the dimensions are sometimes returned to us flipped here, with `rotation` set to 90 degrees.
+            // Swap them back to make sure the image fits nicely in the container. On iOS, the rotation is always undefined, so this does not apply.
+            if ((rotation === 90 || rotation === 270) && imageWidth > imageHeight) {
```
Comment (aldo-expensify) [inline src/components/ImageView/index.native.js:77]: I'm thinking that this `&& imageWidth > imageHeight` is probably not needed. I think a regular camera usually takes 4:3 (width:height) images, and  store rotation information. This condition would make it miss your `if` when it really should enter.

This is a less common case, but I guess we should still handle it correctly if possible. I will try to find such picture and upload it next week.

Comment (aldo-expensify) [reply src/components/ImageView/index.native.js:77]: Try with this image:

https://github.com/recurser/exif-orientation-examples/blob/master/Landscape_8.jpg?raw=true

Removing `&& imageWidth > imageHeight` results in the image using the maximum space

File: src/components/ImageView/index.native.js [hunk 3]
```diff
@@ -66,12 +66,18 @@ class ImageView extends PureComponent {
...
-        ImageSize.getSize(this.props.url).then(({width, height}) => {
+        ImageSize.getSize(this.props.url).then(({width, height, rotation}) => {
             let imageWidth = width;
             let imageHeight = height;
             const containerWidth = Math.round(this.props.windowWidth);
             const containerHeight = Math.round(this.state.containerHeight);
 
+            // On specific Android devices, the dimensions are sometimes returned to us flipped here, with `rotation` set to 90 degrees.
+            // Swap them back to make sure the image fits nicely in the container. On iOS, the rotation is always undefined, so this does not apply.
+            if ((rotation === 90 || rotation === 270)) {
```
Comment (aldo-expensify) [inline src/components/ImageView/index.native.js:77]:
```suggestion
            if (rotation === 90 || rotation === 270) {
```

Remove the extra set of parenthesis and we are done :)

**Review by aldo-expensify — CHANGES_REQUESTED** [review]
Tested fine, just remove those parenthesis
Screenshot_20221220-121303
Screenshot_20221220-121358

**Comment by aldo-expensify** [comment]
@francoisl do you want me to investigate the backend part of this? or do you want to take it?

**Comment by francoisl** [comment]
Hm up to you, if you feel like taking it go ahead, otherwise I can pick it up after finishing a couple other tasks first. Either way, let's do that as a follow-up task - no need to hold on merging this PR imo.

File: src/components/ImageView/index.native.js (hunk 3 above)
Comment (francoisl) [reply src/components/ImageView/index.native.js:77]: Oh man how did I not see this :|

**Comment by aldo-expensify** [comment]
> Hm up to you, if you feel like taking it go ahead, otherwise I can pick it up after finishing a couple other tasks first. Either way, let's do that as a follow-up task - no need to hold on merging this PR imo.

Yes, it should be a follow-up task and no need to hold. I'm ok with leaving it to you if you want, my intention was just to offer my help on the follow-up pr if you didn't want to do it :)

**Review by aldo-expensify — APPROVED** [review]

**Comment by thesahindia** [comment]
EOD for me. Will test in the morning.

**Comment by aldo-expensify** [comment]
I left it unmerged in case you wanted to wait for other reviews

**Review by thesahindia — APPROVED** [review]
LGTM!

**Comment by francoisl** [comment]
Ok, two approvals so I'm going to merge. Thanks for the reviews and tests guys!
